src/client/providers/queryTransformer.test.ts

```diff
@@ -0,0 +1,66 @@
+import * as assert from 'assert'
+import { of } from 'rxjs'
+import { TestScheduler } from 'rxjs/testing'
+import { transformQuery, TransformQuerySignature } from './queryTransformer'
+
+const scheduler = () => new TestScheduler((a, b) => assert.deepStrictEqual(a, b))
+
+const FIXTURE_INPUT = 'foo'
+const FIXTURE_RESULT_MERGED = 'foo bar'
+const FIXTURE_RESULT = 'bar'
+
+describe('transformQuery', () => {
+    describe('0 providers', () => {
+        it('returns original query', () =>
+            scheduler().run(({ cold, expectObservable }) =>
+                expectObservable(
+                    transformQuery(cold<TransformQuerySignature[]>('-a-|', { a: [] }), FIXTURE_INPUT)
+                ).toBe('-a-|', {
+                    a: FIXTURE_INPUT,
+                })
+            ))
+    })
+
+    describe('1 provider', () => {
+        it('returns result from provider', () =>
+            scheduler().run(({ cold, expectObservable }) =>
+                expectObservable(
+                    transformQuery(
+                        cold<TransformQuerySignature[]>('-a-|', {
+                            a: [q => of(FIXTURE_RESULT)],
+                        }),
+                        FIXTURE_INPUT
+                    )
+                ).toBe('-a-|', { a: FIXTURE_RESULT })
+            ))
+    })
+
+    describe('2 providers', () => {
+        it('returns a single query transformed by both providers', () =>
+            scheduler().run(({ cold, expectObservable }) =>
+                expectObservable(
+                    transformQuery(
+                        cold<TransformQuerySignature[]>('-a-|', {
+                            a: [q => of(q), q => of(`${q} ${FIXTURE_RESULT}`)],
+                        }),
+                        FIXTURE_INPUT
+                    )
+                ).toBe('-a-|', { a: FIXTURE_RESULT_MERGED })
+            ))
+    })
+
+    describe('Multiple emissions', () => {
+        it('returns stream of results', () =>
+            scheduler().run(({ cold, expectObservable }) =>
+                expectObservable(
+                    transformQuery(
+                        cold<TransformQuerySignature[]>('-a-b-|', {
+                            a: [q => of(q)],
```

Comment:
If the first query transform is just the identity func, then this doesn't really test the behavior of "all query transform funcs are applied". Instead you should test something like:

- 1st query transform func appends `a` to the string
- 2nd query transform func appends `b` to the string

and test that the result is the original string with `ab` appended.

(I think the current test would pass even if your logic was like `this.providers[1](query)`, which would obviously be incorrect and would not be caught by the test.)